### PR TITLE
Add urls checkup

### LINKED ISSUE #35921: Travis CI should run linkchecker
Would allow us to find these issues before getting reviewer approved:

https://github.com/rust-lang/rust/pull/35845

https://github.com/rust-lang/rust/pull/35786

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes #35921.

### CODE REVIEW

**Comment by frewsxcv** [comment]
```
sh: 1: cd: can't cd to src/src/tools/linkchecker
The command "docker run -v `pwd`:/build rust sh -c " ./configure --llvm-root=/usr/lib/llvm-3.7 && make tidy && make check-notidy -j4 && (cd src/src/tools/linkchecker && cargo run ../../lib*) "" exited with 2.
```

**Comment by GuillaumeGomez** [comment]
Woups! Fixed.

**Comment by GuillaumeGomez** [comment]
I need to include cargo into travis as well. Any preferred way of how I should do it?

**Comment by nikomatsakis** [comment]
@GuillaumeGomez I have no idea :( I know very little about travis setup.

**Comment by nikomatsakis** [comment]
@GuillaumeGomez are you sure that is the problem? Looking at the error message:

```
sh: 1: cd: can't cd to src/src/tools/linkchecker
```

it seems like you might need to remove one `src/` or add an extra `..` or something.

**Comment by GuillaumeGomez** [comment]
That's not what I see:

```bash
sh: 1: cargo: not found
```

**Comment by frewsxcv** [comment]
This is probably not too constructive of a comment, but if Travis CI started using rustbuild, I think we'd have cargo available. From what I understand, with the old `make`-based system (which is what Travis CI is using now), cargo is *not* available.

**Comment by GuillaumeGomez** [comment]
It uses make and docker.

**Comment by GuillaumeGomez** [comment]
cc @sanxiyn

**Comment by sanxiyn** [comment]
@frewsxcv is right. Travis is currently using make-based build and cargo is not available.

**Comment by GuillaumeGomez** [comment]
My question was: how could I add it? Do I download it from the rust-lang website or do I look for another source?

**Comment by nikomatsakis** [comment]
@GuillaumeGomez seems like we'd have to download it -- the `.travis.yml` script exposes the various commands that get executed. I think what @frewsxcv and @sanxiyn were saying (and I may be wrong; they can confirm) is that rustbuild already does this download, so switching to use rust-build would cause cargo to get downloaded (though you'd may still have to run it from the right place, not sure if it would be available on PATH)

**Comment by alexcrichton** [comment]
Ah ok, so looking at this, I think we've got a few options of how to proceed:

* As-is I don't think this PR will accomplish the goal as stated, unfortunately. The `linkchecker` executable expects to be run with the directory of generated docs, so passing `../lib*` I don't think will work. As Travis is saying, though, `cargo` is not an available executable in any case. That is, it'll have to be downloaded from somewhere to run.

* We could also add support to run `linkchecker` as part of the makefile build system. Right now, however, the makefiles don't download Cargo, so that would need to be added as well.

* Finally, we could switch Travis to the rustbuild build system (like @frewsxcv was mentioning). I believe at this point the rustbuild build system now runs strictly more tests than the makefiles (e.g. the linkchecker), so we'll run these tests on Travis simply by using rustbuild. This can be done by just adding `--enable-rustbuild` to the configure line.

My preference and recommendation would be to just switch to using rustbuild. That may, however, expose some bugs that also need to be solved.

Basically the tl;dr; is that I don't believe this will be a trivial patch, but rustbuild has the most promise for being the easiest route.

**Comment by GuillaumeGomez** [comment]
@alexcrichton: Thanks for the propositions! I'll wait for the rustbuild switch and then I'll just run it on the docs.

**Comment by alexcrichton** [comment]
Ok, I'm going to close this in https://github.com/rust-lang/rust/pull/36180 which is attempting to enable rustbuild
